### ISSUE TITLE
:terminal command to jump between prompts

### ISSUE THREAD

**Issue by gerion0** [neovim/neovim#9209]
I think a pretty nice feature of the terminal mode (`:terminal`) would be the possibility to jump between prompts, when being in normal mode in the terminal buffer.

E.g. my window looks like so:
```
+-------------------------------------------------------------------------+
|$ ls                                                                     |
|decode.c  encode.c  executor.c  gc.c  typval.c           typval_encode.h |
|decode.h  encode.h  executor.h  gc.h  typval_encode.c.h  typval.h        |
|$ █                                                                      |
|                                                                         |
|--TERMINAL--                                                             |
+-------------------------------------------------------------------------+
```
and I am in normal mode and type e.g. `:PrevPrompt` and get:
```
+-------------------------------------------------------------------------+
|$ █s                                                                     |
|decode.c  encode.c  executor.c  gc.c  typval.c           typval_encode.h |
|decode.h  encode.h  executor.h  gc.h  typval_encode.c.h  typval.h        |
|$                                                                        |
|                                                                         |
|--TERMINAL--                                                             |
+-------------------------------------------------------------------------+
```
Then I type e.g. `:NextPrompt` and would be at first position again:
```
+-------------------------------------------------------------------------+
|$ ls                                                                     |
|decode.c  encode.c  executor.c  gc.c  typval.c           typval_encode.h |
|decode.h  encode.h  executor.h  gc.h  typval_encode.c.h  typval.h        |
|$ █                                                                      |
|                                                                         |
|--TERMINAL--                                                             |
+-------------------------------------------------------------------------+
```

Of course some can search `$` and jump between prompts but not in a clean way since the output of terminal commands could contain the prompt string (or parts of it).

**Comment by justinmk** [comment]
Could be a reasonable default behavior of `[[` and `]]` in `buftype=terminal` buffers.  Can be implemented as a plugin.

**Comment by bew** [comment]
I'm curious, why do you want such a feature?

**Comment by bew** [comment]
@gerion0 :+1: I've been thinking about something like that too some time ago (but not specifically inside neovim)

@justinmk good luck with that, `$PS1`s can be very complex, and change dynamically.. It'll never work with my prompt ^^
Also, this would depend based on the used shell (sh/fish/zsh/..)

I think this could be implemented as a shell plugin (one for each shell), that would tell the terminal (in this case, the vterm inside neovim) `something` everytime the shell accepts a line (execute a command), then neovim could use this `something` to store the position of the cursor, and offer a way to navigate through those positions..
So a `something`-protocol could be designed, and then implemented by neovim to be able to navigate between prompts.

(just some thoughts :smiley:  )

**Comment by gerion0** [comment]
Unfortunately not. I guess, `$PS1` will not work. E.g. my `$PS1` contains a full Python script, color codes and is very specific to zsh.

I hoped that libvterm provides a way to get the positions, but I don't know if (other) terminals can recognize the prompt.

However, traditional terminal emulators can show the current running program. Maybe something like this would be a valid way:
```
if current_running_program == '$SHELL':
    prompt = current_line_of_cursor()
```
and a list to store all prompts.

**Comment by jamessan** [comment]
Some of this was already discussed in #8805.

@gerion0 I would suggest testing out emacs' terminal mode to see how it works with your prompt, since it is [supposed](https://github.com/neovim/neovim/issues/8805#issuecomment-410110967) to handle navigating between shell prompts.

**Comment by glacambre** [comment]
Hi, if you're using zsh I implemented something that does exactly this! You can find it [here](https://github.com/glacambre/shelley).

It suffers from a bug where your prompt history will be messed up once neovim start dropping lines from your terminal buffer (be sure to `:set scrollline=-1` in order to make it take as long as possible before the bug kicks in) but otherwise I've been using it for more than a year and it's been working fine.
I have no idea how to write plugins and nobody used it besides me so feel free to open issues in my repo if you encounter problems :).

**Comment by jneen** [comment]
It would potentially be possible to integrate with `$PROMPT_COMMAND`, which runs on every prompt - if there was a hook in there that set a mark somehow we could tell where the prompts were even if $PS1 was dynamic.

**Comment by justinmk** [comment]
@jneen Indeed, that should be trivial once we have `$NVIM` available in `:terminal`. I'm going to work on that today.

**Comment by gerion0** [comment]
Just to link it here: Kitty has gained a [really similar feature](https://sw.kovidgoyal.net/kitty/shell-integration/) since quite some time. Kitty recognizes shell prompts with that automagically and allows for example:
- jumping between prompts
- open the last command's output in less
- move the shell cursor with the mouse

Maybe neovim can use the [same scripts](https://github.com/kovidgoyal/kitty/tree/master/shell-integration) to provide similar features or take them as reference.

**Comment by bew** [comment]
It's not just a script, the terminal must support the [semantic prompt escapes sequences (OSC 133)](https://gitlab.freedesktop.org/Per_Bothner/specifications/blob/master/proposals/semantic-prompts.md)

The script in question is simply setting up the shell to send these sequences to the terminal before/after each prompt to mark various zones of the terminal session (prompt, input, output)

**Comment by glacambre** [comment]
Related: https://github.com/neovim/neovim/pull/22159 (support for allowing plugins to create terminal OSC handlers).

**Comment by ColinKennedy** [comment]
@glacambre could you expand on how one might use https://github.com/neovim/neovim/pull/22159 to solve this issue? I'm not sure how one would check if a terminal supports OSC 133 but I did try the latest Alacritty and Wezterm and hoped/expected TermRequest / TermResponse to fire while making new prompt lines but that didn't happen. I don't know much about terminal handling / OSC standards so please forgive the question if it's simple.

**Comment by JohnEricson** [comment]
I noticed that VS Code has this functionality when using the integrated shell. It uses [OSC commands to detect the prompt in the scrollback](https://code.visualstudio.com/docs/terminal/shell-integration#_vs-code-custom-sequences-osc-633-st). I would like to get the terminal in Neovim to work with this behavior in VS Code. Anyone knows how I can send these OSC commands from Neovim?

**Comment by glacambre** [comment]
@ColinKennedy @JohnEricson Here's a proof of concept:
```vim
function! OnTermRequest()
  let kind = v:termrequest[6]
  " 'B' marks the end of a prompt, that's what we want to know about
  if kind != 'B'
    return
  endif
  echom v:termrequest
  if !exists("b:prompts")
    let b:prompts = []
  endif
  " We now need to find the positiong of the prompt in the buffer. We do this
  " by finding the last line of the buffer that isn't empty. The reason we
  " have to do this instead of just using nvim_buf_line_count() is that when
  " the height of a terminal buffer is smaller that the height of a window,
  " the terminal buffer gets filled with empty lines until it matches the
  " height of the window.
  let lines = nvim_buf_get_lines(0, 0, -1, 0)
  let last_non_empty_line = 0
  for linenr in reverse(range(0, len(lines) - 1))
    if lines[linenr] != ""
      let last_non_empty_line = linenr
      break
    endif
  endfor
  call add(b:prompts, last_non_empty_line)
endfunction
au! TermRequest * call OnTermRequest()
function! JumpToPrompt(n)
  let l = len(b:prompts)
  call cursor(b:prompts[(l + a:n) % l] + 1, 1)
endfunction
```
It's not proper OSC 133 support - it's lacking support for multiple OSC commands, will break with fancy prompts and very fast input. But `call JumpToPrompt(n)` should take you to the right prompt assuming you set your PS1 to something simple like `PS1=$'%{\e]133;P;k=i\a%}'>$'%{\e]133;B\a%}'` (zsh) or `PS1='\[\e]133;P;k=i\a\]'>'\[\e]133;B\a\]'` (bash) and do not have a PS2.

It can be made more robust with some effort but cannot be made rock-solid with the current implementation of TermRequest autocommands in Neovim - these autocommands are currently executed asynchronously, meaning that the contents of the buffer might have changed between the moment the TermRequest was emitted by the shell and the moment the autocommand actually gets executed.

**Comment by justinmk** [comment]
https://github.com/neovim/neovim/pull/32736

**Comment by viniarck** [comment]
> [#32736](https://github.com/neovim/neovim/pull/32736)

I'm so glad this has landed, much appreciated @justinmk.

**Comment by justinmk** [comment]
Related proposal: OSC 3008 https://github.com/systemd/systemd/blob/main/docs/OSC-CONTEXT.md
